### PR TITLE
"Read more" for jobs

### PR DESCRIPTION
**Pull request by yaelerdr**
- Added an option to read more about a job
- If entering in the url - /job/idNumber, and idNumber isn't the id of any job -
  added a "no_such_job.html" page.
- For students - have the option to apply for a job. For recruiter - no
- If a student apply - the applicant is saved to db, and the student won't be able to apply again
- Recruiter who is viewing his own job should see a note about it
- Added tests for the job is rendered according to the user is trying to access it

### CODE REVIEW

File: job_application/templates/apply_success.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{% extends 'base.html' %}
+{% block content %}
+    <h1>You have successfully applied for this job !</h1>
+        <a href="/">Go back to finding a job!</a>
```
Comment (dmaizel) [inline job_application/templates/apply_success.html:4]: indentation?

File: jobboard/templates/jobboard/job.html [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            {% else %}
+            <button onclick="location.href = 'apply_for_job/{{ job.id }}'" type="submit">APPLY NOW</button>
+            {% endif %}
+            {% endif %}
+
+<!--            {% if is_recruiter %}-->
+<!--            {% if job_data.recruiter.user.id == user.id %}-->
+<!--            <button onclick="location.href = 'edit_job/{{ job.id }}'" type="submit">Edit this job</button>-->
+<!--            {% endif %}-->
+<!--            {% endif %}-->
```
Comment (dmaizel) [inline jobboard/templates/jobboard/job.html]: Why is this here?

File: jobboard/templates/jobboard/job.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            <h5>{{ job_data.description }}</h5>
+
+            {% if user.is_authenticated %}
+            {% if is_student %}
+            {% if is_student_applied %}
+            <p>YOU HAVE ALREADY APPLIED FOR THIS JOB</p>
+            {% else %}
+            <button onclick="location.href = 'apply_for_job/{{ job.id }}'" type="submit">APPLY NOW</button>
+            {% endif %}
+            {% endif %}
```
Comment (dmaizel) [inline jobboard/templates/jobboard/job.html]: IMO, it is too hard to understand what's going on here without indentation...

File: jobboard/templates/jobboard/job.html (hunk 2 above)
Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html]: Going forward I want to add for recruiter that opens it's own job a button that redirects him to edit his job. This editing page isn't implemented yet. But this will be the logic - so I have commented it.

File: jobboard/templates/jobboard/job.html (hunk 3 above)
Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html]: Yeah, your'e right. not sure why it was uploaded like this.. It wasn't that way on my comp. I'll try to push it again as it's on my comp

**Comment by guyy13** [comment]
Clicking the 'APPLY NOW' button results in a 404 error

File: jobboard/templates/jobboard/job.html [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{% extends 'base.html' %} {% block content %}
+
+<div class="container-fluid">
+    <div class="row">
+        <div class="col-md p-5 text-align">
+            <h3>{{ job_data.title }} at {{ job_data.company }}</h3>
+            <h5>{{ job_data.description }}</h5>
```
Comment (guyy13) [inline jobboard/templates/jobboard/job.html]: I think this page needs to include all of the data about the job, it doesn't make sense that you click on "read more" but then fewer details are visible than before.

**Review by guyy13 — CHANGES_REQUESTED** [review]

File: jobboard/templates/jobboard/job.html (hunk 4 above)
Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html]: Yeah, you are right. 
Will add more details

**Review by guyy13 — CHANGES_REQUESTED** [review]
Ok, now the apply button is working, thought I think you need to add a test in order to check that it's actually working, and creates the relevant Job Application instance in the DB

**Comment by yaelerdr** [comment]
> Ok, now the apply button is working, thought I think you need to add a test in order to check that it's actually working, and creates the relevant Job Application instance in the DB

done

**Review by eldaraharonson — APPROVED** [review]

File: jobboard/views.py [hunk 5]
```diff
@@ -45,3 +46,26 @@ def board(request):
...
+    if context["is_student"]:
+        context["is_student_applied"] = check_if_student_already_applied(Student.get_student(request.user),
+                                                                         context["job_data"])
+    return render(request, "jobboard/job.html", context)
+
+
+def check_if_student_already_applied(student, job):
+    applications_for_current_job = set(Application.get_applications_by_job(job))
+    applications_for_current_student = set(Application.get_applications_by_student(student))
+    return len(applications_for_current_job.intersection(applications_for_current_student)) > 0
```
Comment (guyy13) [inline jobboard/views.py:118]: nice!

**Review by guyy13 — APPROVED** [review]
Nice work Yael!

File: jobboard/templates/jobboard/job.html [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                    {% else %}
+                        <button onclick="location.href = 'apply_for_job/'" type="submit">APPLY NOW</button>
+                    {% endif %}
+                {% endif %}
+
+<!--                {% if is_recruiter %}-->
+<!--                    {% if job_data.recruiter.user.id == user.id %}-->
+<!--                        <button onclick="location.href = 'edit_job/{{ job.id }}/'" type="submit">Edit this job</button>-->
+<!--                    {% endif %}-->
+<!--                {% endif %}-->
```
Comment (YarinBou) [inline jobboard/templates/jobboard/job.html]: why do these lines in the comment?

File: jobboard/tests.py [hunk 7]
```diff
@@ -182,3 +182,97 @@ def test_page_display_to_user_that_is_recruiter(client):
...
+
+
+@pytest.mark.django_db
+def test_try_displaying_non_exist_job_user_is_not_student_and_not_recruiter(client):
+    example_user = User.objects.create_user("username", "password")
+    client.force_login(example_user)
+    # job with id=10 does not exist
+    response = client.get("/job/10/")
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'jobboard/no_such_job.html')
```
Comment (YarinBou) [inline jobboard/tests.py]: super-duper tests! great job Yael.

**Review by YarinBou — APPROVED** [review]

File: jobboard/templates/jobboard/job.html (hunk 6 above)
Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html]: I'm waiting for the edit job page to be added. Afterwards, I'll add here the link to that page, and add tests to the render of that page. I'll push tomorrow a PR without these lines in the meantime

**Review by benmenachem — APPROVED** [review]
well done Yael 
everything looks good

File: jobboard/templates/jobboard/job.html [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+<div class="container-fluid">
+    <div class="row">
+        <div class="col-md p-5 text-align">
+            <h2>{{ job_data.title }} </h2>
+            <h6> Company: {{ job_data.company }}, in {{ job_data.city }}</h6>
+            <h7>{{ job_data.job_type }} job ;</h7>
+            <h7> Work from: {{ job_data.work_from }}</h7>
+            <br> <br>
+            <h6><strong>{{ job_data.description }}</strong></h6>
+            <br>
```
Comment (ifireball) [inline jobboard/templates/jobboard/job.html:12]: 3 nested `if`s without tests...

File: jobboard/tests.py [hunk 9]
```diff
@@ -182,3 +182,97 @@ def test_page_display_to_user_that_is_recruiter(client):
...
+    assert response.context['is_student'] is True
+    assert response.context['is_recruiter'] is False
+    # example_student_a haven't applied to job with id=1
+    assert response.context["is_student_applied"] is False
+    job_to_display = Job.objects.get(id=1)
+    assert response.context['job_data'] == job_to_display
+
+
+@pytest.mark.django_db
+def test_apply_button_for_student_trying_to_apply_and_test_after_apply(client, example_student_a):
```
Comment (ifireball) [inline jobboard/tests.py]: I think this test should probably be broken apart to multiple smaller tests. Usually you want to have only one HTTP requests per-test unless those requests are very tightly coupled.

**Review by ifireball — CHANGES_REQUESTED** [review]

File: jobboard/tests.py (hunk 9 above)
Comment (yaelerdr) [reply jobboard/tests.py:202]: OK, no problem.  I'll split it

File: jobboard/templates/jobboard/job.html (hunk 8 above)
Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html:12]: will add tests for it

Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html:12]: @ifireball Should I add separate tests for different users for this one?

Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html:12]: @ifireball Should I add separate tests for different users for this one?

Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html:12]: @ifireball I checked the content of the `job_data` , `is_student` and `is_recruiter`  keys in the dictionary in each of the tests.

how else can I check those ifs?

File: jobboard/tests.py (hunk 9 above)
Comment (yaelerdr) [reply jobboard/tests.py]: @ifireball  I think here, you can call these requests tightly coupled.. I'm checking for the same user that he haven't applied for a job - checking the `is_student_applied = False`; Then, taking this same student to the page of applying to a job, checking if `is_student_applied = True` , and then, making this same student to manually trying to access the page of applying the same job again. 
 I don't want to save the application of that student to the DB, so I think its better all will be in one test.

Comment (ifireball) [reply jobboard/tests.py]: I think the code would still be easier to understand if those are separate tests. And its not necessary in practice that the HTTP calls you test will happen in the order you test them in.

Comment (ifireball) [reply jobboard/tests.py]: For HTTP calls to be coupled you need to have some information sharing mechanism to be passed between them like e.g. a cookie, you don't have that here.

Comment (yaelerdr) [reply jobboard/tests.py]: OK. I'll change it

File: jobboard/templates/jobboard/job.html (hunk 8 above)
Comment (guyy13) [reply jobboard/templates/jobboard/job.html:12]: I think what @ifireball means is that you should take the logic out of the template.
You can do something similar to what @eldaraharonson did with the board view on 'jobboard/views.py', where he created functions that set the content in the context dictionary according to the user who sent the request.
Then you can test it by checking that the correct data is passed through the context.
You can also look at what I did with the nav bar links it follows the same idea

Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html:12]: Eldar did the exact same thing as I did, just that he made separate functions that insert the content to dictionary. and he used the `user_is_student` and `user_is_recruiter` in his html page. And - I did exactly what you did, same functions implementation, just inside the `detail_view` function.
I check in each one of the tests these fields. I think it's exactly the same

Comment (guyy13) [reply jobboard/templates/jobboard/job.html:12]: Yea actually you're right Eldar did the same thing.
What I did with the nav bar is handling all the `ifs` outside but I'm not sure how and if you can do that here as in one case you have to display a button and another case a paragraph.
Anyone else has an idea?

Comment (ifireball) [reply jobboard/templates/jobboard/job.html:12]: Here is one way to to it that I described during the demo on Sunday:

In the view code:
```python
if Student.is_student(request.user.id):
    user_indicator_template = "student_user_indicator.html"
elif Recruiter.is_recruiter(request.user.id):
    user_indicator_template = "requiter_user_indicator.html"
else:
    user_indicator_template = "other_user_indicator.html"
context["user_indicator_template"] = user_indicator_template
```

In the template code:
```jinja2
{% include user_indicator_template %}
```

File: jobboard/templates/jobboard/job.html (hunk 6 above)
Comment (benmenachem) [reply jobboard/templates/jobboard/job.html]: @yaelerdr the edit job page won't make it. so I think you could update it in your PR. we have 9 PR's open so I don't think we will be able to add more features before the final demo

File: jobboard/templates/jobboard/job.html (hunk 8 above)
Comment (yaelerdr) [reply jobboard/templates/jobboard/job.html:12]: @ifireball  fixed that. Please take a look

Comment (guyy13) [reply jobboard/templates/jobboard/job.html:12]: looks good yael

**Comment by guyy13** [comment]
I went over the PR again, besides the conflicts, I think it looks good.

**Comment by guy9050** [comment]
please resolve the conflicts

**Comment by yaelerdr** [comment]
> please resolve the conflicts

@guy9050  I have solved them ~2 hours ago. It's because there were other merged PR a minute ago. I am solving it now and will push it ASAP

**Comment by yaelerdr** [comment]
> > please resolve the conflicts
> 
> @guy9050 I have solved them ~2 hours ago. It's because there were other merged PRs. I am solving it now and will push it ASAP

> please resolve the conflicts

@guy9050 Have solved the conflicts. Please take a look

**Review by YarinBou — APPROVED** [review]

File: SJMaster/urls.py [hunk 10]
```diff
@@ -7,4 +7,5 @@
     path('', include("recruiter.urls"), name="recruiter_urls"),
     path('', include("login.urls"), name="login_urls"),
     path('', include("jobboard.urls"), name="jobboard_urls"),
+
```
Comment (guy9050) [inline SJMaster/urls.py]: why this blank line?

File: job_application/views.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
+from datetime import datetime
+from django.shortcuts import render, get_object_or_404
+from jobboard.models import Job
+from jobboard.views import check_if_student_already_applied
+from .models import Application
+from student.models import Student
+
+
+def apply_success(request):
+    return render(request, 'apply_success.html')
```
Comment (guy9050) [inline job_application/views.py]: I can't find where you're using this function. If it's not in use than delete it

**Review by guy9050 — COMMENTED** [review]
just two small things, otherwise lgtm

File: SJMaster/urls.py (hunk 10 above)
Comment (yaelerdr) [reply SJMaster/urls.py]: For nothing. Will be deleted

File: job_application/views.py (hunk 11 above)
Comment (yaelerdr) [reply job_application/views.py]: I think you are right. Will be next to the computer on 14:00, and will delete it

**Review by ifireball — APPROVED** [review]
